### ISSUE TITLE
Add a new symbols function with well defined types

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#27354]
There was previous discussion about this in the PR gh-23181 (CC @eagleoflqj). I also have a related PR gh-27353.

The `symbols` function is impossible to use for downstream code that wants to use static typing. There are three basic problems that prevent it from working properly:

Firstly symbols returns different types depending on the value (rather than the type) of the string argument:
```python
In [1]: type(symbols('x, y'))
Out[1]: tuple

In [2]: type(symbols('x'))
Out[2]: sympy.core.symbol.Symbol
```
This makes it impossible to infer the types in the most common usage of the `symbols` function. The reason this is done is so that you can do:
```python
x = symbols('x')
x, y = symbols('x, y')
```
Otherwise you would need something like this for a single symbol:
```python
[x] = symbols('x')
x, = symbols('x')
```
Of course there is no reason why you need to use `symbols` for a single symbol but that usage is widespread including in the sympy codebase itself.

The second problem with symbols is that it can be used to create other types e.g.:
```python
f, g, h = symbols('f, g, h', cls=Function)
```
In principle this should be understandable to a type checker and I can write type annotations that pyright will understand but I have not succeeded in getting anything that mypy will understand if there is a default value `cls=Symbol`. Trying to make this work would be quite disruptive to users because it would end up meaning that they get errors from mypy any time they use the symbols function.

The third problem is just that it does all kinds of weird stuff like:
```python
In [7]: symbols(['x:3', 'y', 'z', [[['Q:2:2', 'z']]]])
Out[7]: [(x₀, x₁, x₂), y, z, [[[(Q₀₀, Q₀₁, Q₁₀, Q₁₁), z]]]]
```

In the PR gh-27353 I have added an overload that says that when using the parameter `seq=True` the return type will always be a tuple of Symbol:
```python
x, y, z = symbols('x, y, z', seq=True)
```
That makes is possible for both pyright and mypy to infer the types and can be used in downstream codebases that depend on sympy since it works with older sympy versions and can be understood by a type checker. I also used that in the sympy.abc module so that the symbols there have inferred types.

Using `seq=True` like this solves problem 1 because the type is unambiguous but it ignores problem 2 by assuming that it is always Symbol and it also ignores the weird types in problem three but I think that is okay. So this isn't ideal but I think it is good as a starting point that there is at least some well-typed way of calling `symbols` for the most common usage.

The main problems with the `seq=True` approach are:

1. I don't think we really want users to have to write `seq=True` although I think some will be happy to use it if it makes type checking possible.
2. It still assumes that it is always `Symbol` so e.g. `symbols('f, g, h', cls=Function)` would be interpreted incorrectly by a type checker.

I think that the proper solution is just that we need a new function that is suitable for anyone downstream to create symbols of different types in a statically typed context. I have a demonstration here of what I have done with the overloads in gh-27353 as well as adding a new function called `symbols_type` which can be called like `symbols_type('x, y, z', Function)`:
```python
from typing import (
    reveal_type,
    TypeVar,
    Callable,
    Any,
    ParamSpec,
    Concatenate,
    overload,
    Literal,
)

from collections.abc import Iterable
from sympy import Expr, Symbol, Function, MatrixSymbol

@overload
def symbols(names: str, *, cls: Any = Symbol, seq: Literal[False] = False,
    **kwargs: bool | int,) -> Any: ...
@overload
def symbols(names: str, *, cls: Any = Symbol, seq: Literal[True],
    **kwargs: bool | int,) -> tuple[Symbol, ...]: ...

def symbols(names: str, *, cls: Any = Symbol, seq=True, **kwargs) -> Any | tuple[Symbol, ...]:
    syms = tuple(cls(name, **kwargs) for name in names.split(", "))
    if not seq and len(syms) == 1:
        return syms[0]
    else:
        return syms


T = TypeVar("T")
P = ParamSpec("P")


def symbols_type(
    names: str | Iterable[str | tuple[str, int] | tuple[str, int, int]],
    func: Callable[Concatenate[str, P], T],
    *args: P.args,
    **kwargs: P.kwargs,
) -> tuple[T, ...]:
    """Make symbols of type T.

    >>> symbols_type('x, y, z', Symbol)
    (x, y, z)
    >>> symbols_type(['x', 'y', 'z'], Symbol)
    (x, y, z)

    Symbols of different types can be made:

    >>> symbols_type('f, g, h', Function)
    (f, g, h)
    >>> symbols_type('A, B, C', MatrixSymbol, 2, 2)
    (A, B, C)

    Keyword arguments such as assumptions for Symbol can be passed:

    >>> symbols_type('a, b, c', Symbol, integer=True)
    (a, b, c)

    Multiple symbols can be generated with numbered names:

    >>> symbols_type([('x', 3), ('y', 2)], Symbol)
    (x0, x1, x2, y0, y1)
    >>> symbols_type([('x', 3, 6)], Symbol)
    (x3, x4, x5)
    """

    match names:
        case str():
            names_list = [n.strip() for n in names.split(",") for n in n.split()]
        case Iterable():
            names_list = [n for item in names for n in _parse_item(item)]
        case _:
            raise TypeError("Expecting str or iterable")

    return tuple(func(name, *args, **kwargs) for name in names_list)


def _parse_item(item: str | tuple[str, int] | tuple[str, int, int]) -> list[str]:
    """Handle e.g. ('x', 3) -> ['x0', 'x1', 'x2']"""
    match item:
        case str(name):
            return [name]
        case (str(name), int(stop)):
            return [name + str(i) for i in range(stop)]
        case (str(name), int(start), int(stop)):
            return [name + str(i) for i in range(start, stop)]
        case _:
            raise TypeError("Expected str, (str,int), or (str,int,int)")


a, b, c = symbols("a, b, c")
x, y, z = symbols("x, y, z", seq=True)
i, j, k = symbols_type("i, j, k", Symbol)
f, g, h = symbols_type("f, g, h", Function)
A, B, C = symbols_type("A, B, C", MatrixSymbol, 2, 2)

reveal_type(a)  # Any
reveal_type(x)  # Symbol
reveal_type(i)  # Symbol
reveal_type(f)  # type[AppliedUndef]
reveal_type(f(x))  # Expr
reveal_type(A)  # MatrixSymbol


if __name__ == "__main__":
    import doctest

    doctest.testmod()
```
My suggestion would be to leave the existing `symbols` function as is apart from adding hints for `seq=True` and then add a new function suitable for downstream use and also that should be mostly used throughout the sympy codebase. Here I have suggested calling the function `symbols_type` because the idea is that you pass the type of symbol that you want as the second argument. Other names like `syms` etc could be used.

Using positional arguments for the type means we can pass additional positional arguments to the symbol constructor e.g.:
```python
In [2]: A, B, C = symbols_type("A, B, C", MatrixSymbol, 2, 2)

In [3]: A.shape
Out[3]: (2, 2)
```
This is something that needs a `lambda` with `symbols`:
```python
In [5]: A, B, C = symbols("A, B, C", cls=lambda n: MatrixSymbol(n, 2, 2))
```
I don't think that it is possible with `symbols_type` to have a default argument for the type i.e. to have it default to `Symbol` without complaints from mypy. Maybe it is possible and someone else can figure that out but it is really just a deficiency in mypy because there are many ways to get pyright to understand this. For now though we could add something like `symbols_type` that has a non-optional second argument so to use it with Symbol it would be `symbols_type('x, y, z', Symbol)` or perhaps `syms('x, y, z', Symbol)`.

The implementation that I showed allows creating multiple symbols using range syntax like `symbols_type([('x', 10)], Symbol)`. I think that is better than `symbols('x:10')`  because often I want to create `n` symbols where `n` is a Python variable and it is awkward to format that into a string. I don't think that it is worth supporting any of the other cases that `symbols` handles. ultimately in more complicated cases you should just use a loop or something because the purpose of symbols is just to be convenient for the common case where all the names are spelled out.

Obviously there are many possibilities for what the name could be and what sort of options and arguments it should take but I think that many users are going to want a function that works with type checkers so we need to add something.

**Comment by bjodah** [comment]
I have always shared the frustration with `symbols('x')` not returning a tuple, so much so that I often write my own `symbols` function for my projects. I agree that backwards comp. makes any fundamental changes to `symbols` a no go. So I am in favor of adding a new funtion, I'm fine with "symbols_type" as a name.

**Comment by oscarbenjamin** [comment]
> For now though we could add something like `symbols_type` that has a non-optional second argument so to use it with Symbol it would be `symbols_type('x, y, z', Symbol)` or perhaps `syms('x, y, z', Symbol)`.

Maybe `syms` could be a shortcut for `symbols_type('x, y, z', Symbol)`. Then you could do:
```python
x, y, z = syms('x, y, z')
f, g, h = symbols_type('f, g, h', Function)
```
Then that would suggest that `syms_type` would be a better name than `symbols_type` for consistency though:
```python
x, y, z = syms('x, y, z')
f, g, h = syms_type('f, g, h', Function)
```

**Comment by sylee957** [comment]
My opinion is that you can simply use `map(Symbol, ['x', 'y', 'z'])` or any basic looping/repeating the code is sufficient for this though.
So we don't have to do extra work for anything to be useful.

I think that the only reason that `symbols` is useful is that it can parse the string ("x, y, z" or "x y z")
However, I don't think that parsing things like that is useful, because they are too basic like children.

> I have always shared the frustration with symbols('x') not returning a tuple, so much so that I often write my own symbols function for my projects

One of my coding style I liked to suggest (for example, in the tutorial session)
is that use you should just distinguish usage of `Symbol` for creating single symbol, 
and use `symbols` for creating multiple symbols,
If you don't use use something like `symbols('x')` though, maybe it can make you do less mistakes like that though.

The other thing I discover is that SymEngine may not have `symbols` at all, because with more strict typing of C++,
they may already have concluded that having `symbols` is not worth, and it is only implemented directly Python wrapper

https://github.com/symengine/symengine.py/blob/bf8f69a68d168ca3b764879387951c618f5d371e/symengine/utilities.py#L11-L180

**Comment by oscarbenjamin** [comment]
I think there is something valuable about the `symbols` function. It would be great if there was some way like a macro that could make it possible to avoid duplicating the symbol names and variable names like in Matlab or Julia:
```
@syms x, y, z

eq = x**2 - y
```
Short of that I think that the symbols function is the most ergonomic way to do something that is very common in code that uses sympy which is just to declare a bunch of symbols as local variables. Since the names have to be duplicated it is useful that you can see the parallel with the commas to minimise the chance of an error:
```python
a, b, c, d, e = symbols('a, b, c, d, e')
```
When you need a lot of symbols and some people do need a lot then I think it is not really possible to do better than this.

**Comment by sylee957** [comment]
> It would be great if there was some way like a macro that could make it possible to avoid duplicating the symbol names and variable names like in Matlab or Julia:

Specifically, I also think that `symbols` has no problem for dynamic typing.
However, for using static typing with it, I don't think that there would be obvious and nice suggestions though.
The suggestions to use options (like `seq=True`) or either having new almost duplicate function (let's say, `symbols_list`) just raises redundancy and defeats the point to use `symbols` for conciseness for me.

**Comment by asmeurer** [comment]
Is the fact that the length of the tuple returned by symbols would still be dependent on the value an issue for typing?

**Comment by asmeurer** [comment]
Although it really feels like in principle a static typing extension should be able to parse string literal inputs to `symbols` and infer types from it. I don't know if existing Python type checkers support this sort of thing.

**Comment by oscarbenjamin** [comment]
> Is the fact that the length of the tuple returned by symbols would still be dependent on the value an issue for typing?

No, that is okay. There are different fixed length tuple types like `tuple[Symbol]`, `tuple[Symbol, Symbol]` etc but also a variable length one `tuple[Symbol, ...]`. Type checkers assume that much like if it was a list `a, b, c = things` would have the right length.

**Comment by oscarbenjamin** [comment]
> Although it really feels like in principle a static typing extension should be able to parse string literal inputs to `symbols` and infer types from it.

You can make mypy plugins but we are not going to do that and no one would use it anyway. It would be much easier just to make a new symbols function.

**Comment by bjodah** [comment]
> ... minimise the chance of an error:
> 
> a, b, c, d, e = symbols('a, b, c, d, e')
> 
I agree, I once wrote [this abomination](https://github.com/bjodah/pydollar) as an experiment just for this use case. But it doesn't work interactively unless you hack the python interpreter itself.

And without messing with the parsing of python syntax, I think the closest thing we can get is something with metaclasses/and or descriptors.

**Comment by asmeurer** [comment]
> It would be great if there was some way like a macro that could make it possible to avoid duplicating the symbol names and variable names like in Matlab or Julia:

To be fair that already exists, the `var` function.

```py
>>> var('x y z') # x, y, and z are now defined
```

It also uses a hack but not nearly as hacky as adding new syntax (just injecting names into globals; which static checkers do not like).
